### PR TITLE
[Bugfix] Revert to aioprometheus to avoid 307 redirect

### PR DESCRIPTION
**Pull request by robertgshaw2-redhat**
Reverts prometheus logging from `prometheus_client` back to `aioprometheus` to avoid `307 Temporary Redirect`

FIX https://github.com/vllm-project/vllm/pull/2730#issuecomment-2078805215

At current, we get:
```bash
INFO:     Started server process [1620866]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO:     127.0.0.1:34444 - "GET /metrics HTTP/1.1" 307 Temporary Redirect
INFO:     127.0.0.1:34444 - "GET /metrics/ HTTP/1.1" 200 OK
```

Note: this PR breaks some existing functionality 
- I needed to remove logging `cache_info`
- Some of the metrics exposed by default for `prometheus_client` are gone

I also tweaked the name of a label (nit from prior PR). We did not publish a release with that label (merged over the weekend), so I think this is okay

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/engine/metrics.py [hunk 1]
```diff
@@ -15,72 +14,51 @@
...
 # The begin-* and end* here are used by the documentation generator
 # to extract the metrics definitions.
 
-
 # begin-metrics-definitions
 class Metrics:
     labelname_finish_reason = "finished_reason"
 
-    def __init__(self, labelnames: List[str], max_model_len: int):
-        # Unregister any existing vLLM collectors
```
Comment (robertgshaw2-neuralmagic) [inline vllm/engine/metrics.py:29]: This was a hack to enable our CI to pass before

`REGISTRY` is a global value. We were touching internal state since our CI runs `LLM` twice

I just updated the CI to run with `--forked` which obviate the need for this hack

File: vllm/engine/metrics.py [hunk 2]
```diff
@@ -15,72 +14,51 @@
 
 logger = init_logger(__name__)
 
-disable_created_metrics()
```
Comment (robertgshaw2-neuralmagic) [inline vllm/engine/metrics.py:18]: I do not know what this did, but the equivalent does not seem to exist for `aioprometheus`

File: vllm/engine/metrics.py [hunk 3]
```diff
@@ -15,72 +14,51 @@
...
     labelname_finish_reason = "finished_reason"
 
-    def __init__(self, labelnames: List[str], max_model_len: int):
-        # Unregister any existing vLLM collectors
-        for collector in list(REGISTRY._collector_to_names):
-            if hasattr(collector, "_name") and "vllm" in collector._name:
-                REGISTRY.unregister(collector)
-
-        # Config Information
-        self.info_cache_config = Info(
```
Comment (robertgshaw2-neuralmagic) [inline vllm/engine/metrics.py:35]: This had to be removed since `Info` does not exist in `aioprometheus` AFAIK

File: vllm/engine/metrics.py (hunk 2 above)
Comment (robertgshaw2-neuralmagic) [reply vllm/engine/metrics.py:18]: https://github.com/prometheus/client_python/blob/master/prometheus_client/metrics.py#L74

**Review by simon-mo — APPROVED** [review]

File: vllm/engine/metrics.py (hunk 1 above)
Comment (rkooo567) [reply vllm/engine/metrics.py:29]: We had problems using --forked before due to cuda issue (and we removed it). https://github.com/vllm-project/vllm/pull/3631

Seems like metrics test also use the vllm instances, and I am a little concern it will introduce the similar issue again. Why don't we just keep this hack instead?

**Review by rkooo567 — COMMENTED** [review]
LGTM. Just one comment related to --forked usage..

**Comment by hmellor** [comment]
The redirect, while annoying, was technically harmless.

If possible I'd like to keep using `prometheus_client` as it's still actively maintained. As mentioned in https://github.com/prometheus/client_python/issues/1016, the issue is that `prometheus_client` mounts a sub-application whereas `aioprometheus` adds a route for metrics. 

Note that the `app.route` method from `starlette` which `aioprometheus` uses is discouraged by `starlette` itself (see https://github.com/encode/starlette/blob/9f16bf5c25e126200701f6e04330864f4a91a898/starlette/routing.py#L849-L857).

One simple solution that (almost) works is to add the metrics method from `prometheus_client` as a route:

```python
metrics_app = make_asgi_app()
@app.get("/metrics")
async def metrics(request: Request):
    return await metrics_app(
        scope=request.scope,
        receive=request.receive,
        send=request._send
    )
```

Could we consider alternative approaches before reverting and losing some of the functionality such as the info metrics?

File: vllm/engine/metrics.py (hunk 2 above)
Comment (hmellor) [reply vllm/engine/metrics.py:18]: https://prometheus.github.io/client_python/instrumenting/#disabling-_created-metrics

File: vllm/engine/metrics.py (hunk 1 above)
Comment (robertgshaw2-neuralmagic) [reply vllm/engine/metrics.py:29]: As long as the test does not import torch, —forked is fine

**Comment by hmellor** [comment]
(N.B. I've edited my previous comment so it you're only looking at this via email you won't have the latest update)

**Comment by robertgshaw2-neuralmagic** [comment]
@hmellor is there any performance implications to matching a regex?

Forgive me, my understanding of FastAPI is limited

**Comment by hmellor** [comment]
> is there any performance implications to matching a regex?

I don't think so.

> Forgive me, my understanding of FastAPI is limited

I'll explain my solution a little more, I'm also learning as I go. `app.mount` calls the following method in `starlette` ([source](https://github.com/encode/starlette/blob/9f16bf5c25e126200701f6e04330864f4a91a898/starlette/routing.py#L811-L815)):

```python
    def mount(
        self, path: str, app: ASGIApp, name: str | None = None
    ) -> None:  # pragma: nocover
        route = Mount(path, app=app, name=name)
        self.routes.append(route)
```

So, I am bringing the creation and appending of `route` into `api_server.py`. No real changes so far.

The `Mount` class has an attribute `path_regex` ([source](https://github.com/encode/starlette/blob/9f16bf5c25e126200701f6e04330864f4a91a898/starlette/routing.py#L410-L412)) which it generates by calling `compile_path` ([source](https://github.com/encode/starlette/blob/9f16bf5c25e126200701f6e04330864f4a91a898/starlette/routing.py#L125-L180)). When the `path` is `/metrics`, the generated `path_regex` is `^/metrics/(?P<path>.*)$`.

This regex is used in `Mount.match` ([source](https://github.com/encode/starlette/blob/9f16bf5c25e126200701f6e04330864f4a91a898/starlette/routing.py#L418-L449)) to determine if the path belongs to this app. Therefore, the URL must contain `/metrics/` to match, meaning that `/metrics` won't match and results in the reported redirect error.

By changing the `path_regex` to `^/metrics(?P<path>.*)$` we allow `/metrics` to match, and the redirect error is gone.

In conclusion, this change is only to the value of a regex pattern and does not modify the execution of any functions in `fastapi` or `starlette`, so the performance should be unchanged.

**Comment by robertgshaw2-neuralmagic** [comment]
Can you make a PR?
